rosbag2/test/rosbag2/rosbag2_write_integration_fixture.hpp

```diff
@@ -91,20 +87,50 @@ class RosBag2IntegrationTestFixture : public Rosbag2TestFixture
     std::shared_ptr<rosbag2_storage::SerializedBagMessage> message,
     std::string topic_name, size_t number_expected_messages)
   {
-    messages_stored_counter_.insert({topic_name, 0});
     publisher_futures_.push_back(std::async(
         std::launch::async, [this, message, topic_name, number_expected_messages]() {
           auto publisher = publisher_node_->create_publisher<std_msgs::msg::String>(topic_name);
 
-          while (rclcpp::ok() && messages_stored_counter_[topic_name] < number_expected_messages) {
-            publisher->publish(message->serialized_data);
-            // We need to wait a bit, in order for record to process the messages
+          rosbag2_storage_plugins::SqliteWrapper
+          db(database_name_, rosbag2_storage::storage_interfaces::IOFlag::READ_ONLY);
+          while (rclcpp::ok() && count_stored_messages(db, topic_name) < number_expected_messages) {
+            publisher->publish(message->serialized_data.get());
+            // rate limiting
             std::this_thread::sleep_for(50ms);
           }
         }
     ));
   }
 
+  size_t count_stored_messages(rosbag2_storage_plugins::SqliteWrapper & db, std::string topic_name)
+  {
+    // protect against concurrent writes (from recording) that may make the count query throw.
```

Comment:
if called concurrently, shouldn't this be locked then?

Reply:
We decided no for the following reason: the `rosbag2_storage_plugins::SqliteWrapper` is not designed to handle db locking more gracefully than throwing an exception as for the "record" and "play" usecases there will be no concurrent usage of the `SqliteWrapper`. Even for a later "convert" usage concurrent usage of the same db is (almost certain) not needed. So the concurrent use is only needed in tests the handling of db locks is put in the test fixture.
Retry is the most common strategy for dealing with failures of obtaining a db lock. Every other strategy would yield much more complicated code and the test runners timeout protects against infinite retrying.